File: src/quest_editor/model/QuestEntityModel.ts
import {
    entity_data,
    EntityType,
    get_entity_type,
    QuestEntity,
} from "../../core/data_formats/parsing/quest/Quest";
import { Property } from "../../core/observable/property/Property";
import { list_property, property } from "../../core/observable";
import { WritableProperty } from "../../core/observable/property/WritableProperty";
import { SectionModel } from "./SectionModel";
import { Euler, Quaternion, Vector3 } from "three";
import { floor_mod } from "../../core/math";
import { euler, euler_from_quat } from "./euler";
import { vec3_to_threejs } from "../../core/rendering/conversion";
import { Vec3 } from "../../core/data_formats/vector";
import { WritableListProperty } from "../../core/observable/property/list/WritableListProperty";
import { QuestEntityPropModel } from "./QuestEntityPropModel";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { ObjectType } from "../../core/data_formats/parsing/quest/object_types";

// These quaternions are used as temporary variables to avoid memory allocation.
const q1 = new Quaternion();
const q2 = new Quaternion();

export abstract class QuestEntityModel<
    Type extends EntityType = EntityType,
    Entity extends QuestEntity = QuestEntity
> {
    private readonly _model: WritableProperty<number | undefined>;
    private readonly _section_id: WritableProperty<number>;
    private readonly _section: WritableProperty<SectionModel | undefined> = property(undefined);
    private readonly _position: WritableProperty<Vector3>;
    private readonly _world_position: WritableProperty<Vector3>;
    private readonly _rotation: WritableProperty<Euler>;
    private readonly _world_rotation: WritableProperty<Euler>;
    private readonly _props: WritableListProperty<QuestEntityPropModel>;

    /**
     * Don't modify the underlying entity directly because most of those modifications will not be
     * reflected in this model's properties.
     */
    readonly entity: Entity;

    abstract readonly type: Type;

    readonly model: Property<number | undefined>;

    get area_id(): number {
        return this.entity.area_id;
    }

    readonly section_id: Property<number>;

    readonly section: Property<SectionModel | undefined>;

    /**
     * Section-relative position
     */
    readonly position: Property<Vector3>;

    /**
     * World position
     */
    readonly world_position: Property<Vector3>;

    readonly rotation: Property<Euler>;

    readonly world_rotation: Property<Euler>;

    readonly props: ListProperty<QuestEntityPropModel>;

    protected constructor(entity: Entity) {
        this.entity = entity;

        this._model = property(this.get_entity_model());
        this.model = this._model;

        this.section = this._section;

        this._section_id = property(this.get_entity_section_id());
        this.section_id = this._section_id;

        const position = vec3_to_threejs(this.get_entity_position());

        this._position = property(position);
        this.position = this._position;

        this._world_position = property(position);
        this.world_position = this._world_position;

        const { x: rot_x, y: rot_y, z: rot_z } = this.get_entity_rotation();
        const rotation = euler(rot_x, rot_y, rot_z);

        this._rotation = property(rotation);
        this.rotation = this._rotation;

        this._world_rotation = property(rotation);
        this.world_rotation = this._world_rotation;

        this._props = list_property(
            undefined,
            ...entity_data(get_entity_type(entity)).properties.map(
                p => new QuestEntityPropModel(this, p),
            ),
        );
        this.props = this._props;
    }

    set_model(model: number, propagate_to_props: boolean = true): this {
        this._model.val = model;

        if (propagate_to_props) {
            let props: QuestEntityPropModel[];

            switch (this.type) {
                case ObjectType.Probe:
                    props = this.props.val.filter(p => p.offset === 40);
                    break;

                case ObjectType.Saw:
                case ObjectType.LaserDetect:
                    props = this.props.val.filter(p => p.offset === 48);
                    break;

                case ObjectType.Sonic:
                case ObjectType.LittleCryotube:
                case ObjectType.Cactus:
                case ObjectType.BigBrownRock:
                case ObjectType.BigBlackRocks:
                case ObjectType.BeeHive:
                    props = this.props.val.filter(p => p.offset === 52);
                    break;

                case ObjectType.ForestConsole:
                    props = this.props.val.filter(p => p.offset === 56);
                    break;

                case ObjectType.PrincipalWarp:
                case ObjectType.LaserFence:
                case ObjectType.LaserSquareFence:
                case ObjectType.LaserFenceEx:
                case ObjectType.LaserSquareFenceEx:
                    props = this.props.val.filter(p => p.offset === 60);
                    break;

                default:
                    return this;
            }

            for (const prop of props) {
                prop.set_value(model, false);
            }
        }

        return this;
    }

    set_section(section: SectionModel): this {
        if (section.area_variant.area.id !== this.area_id) {
            throw new Error(`Quest entities can't be moved across areas.`);
        }

        this.set_entity_section_id(section.id);

        this._section.val = section;
        this._section_id.val = section.id;

        this.set_position(this.position.val);
        this.set_rotation(this.rotation.val);

        return this;
    }

    set_position(pos: Vector3): this {
        this.set_entity_position(pos);

        this._position.val = pos;

        const section = this.section.val;

        this._world_position.val = section
            ? pos.clone().applyEuler(section.rotation).add(section.position)
            : pos;

        return this;
    }

    set_world_position(pos: Vector3): this {
        this._world_position.val = pos;

        const section = this.section.val;

        const rel_pos = section
            ? pos.clone().sub(section.position).applyEuler(section.inverse_rotation)
            : pos;

        this.set_entity_position(rel_pos);
        this._position.val = rel_pos;

        return this;
    }

    set_rotation(rot: Euler): this {
        floor_mod_euler(rot);

        this.set_entity_rotation(rot);

        this._rotation.val = rot;

        const section = this.section.val;

        if (section) {
            q1.setFromEuler(rot);
            q2.setFromEuler(section.rotation);
            this._world_rotation.val = floor_mod_euler(euler_from_quat(q1.multiply(q2)));
        } else {
            this._world_rotation.val = rot;
        }

        return this;
    }

    set_world_rotation(rot: Euler): this {
        floor_mod_euler(rot);

        this._world_rotation.val = rot;

        const section = this.section.val;

        let rel_rot: Euler;

        if (section) {
            q1.setFromEuler(rot);
            q2.setFromEuler(section.rotation);
            q2.inverse();

            rel_rot = floor_mod_euler(euler_from_quat(q1.multiply(q2)));
        } else {
            rel_rot = rot;
        }

        this.set_entity_rotation(rel_rot);
        this._rotation.val = rel_rot;

        return this;
    }

    protected abstract get_entity_model(): number | undefined;

    protected abstract get_entity_section_id(): number;
    protected abstract set_entity_section_id(section_id: number): void;

    protected abstract get_entity_position(): Vec3;
    protected abstract set_entity_position(position: Vec3): void;

    protected abstract get_entity_rotation(): Vec3;
    protected abstract set_entity_rotation(rotation: Vec3): void;
}

function floor_mod_euler(euler: Euler): Euler {
    return euler.set(
        floor_mod(euler.x, 2 * Math.PI),
        floor_mod(euler.y, 2 * Math.PI),
        floor_mod(euler.z, 2 * Math.PI),
    );
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/models/QuestEntityModel.kt
package world.phantasmal.web.questEditor.models

import world.phantasmal.core.math.floorMod
import world.phantasmal.lib.fileFormats.quest.EntityType
import world.phantasmal.lib.fileFormats.quest.QuestEntity
import world.phantasmal.observable.value.Val
import world.phantasmal.observable.value.list.ListVal
import world.phantasmal.observable.value.list.listVal
import world.phantasmal.observable.value.mutableVal
import world.phantasmal.web.core.minus
import world.phantasmal.web.core.rendering.conversion.vec3ToEuler
import world.phantasmal.web.core.rendering.conversion.vec3ToThree
import world.phantasmal.web.core.timesAssign
import world.phantasmal.web.core.toEuler
import world.phantasmal.web.externals.three.Euler
import world.phantasmal.web.externals.three.Quaternion
import world.phantasmal.web.externals.three.Vector3
import kotlin.math.PI

abstract class QuestEntityModel<Type : EntityType, Entity : QuestEntity<Type>>(
    /**
     * Don't modify the underlying entity directly because most of those modifications will not be
     * reflected in this model's properties.
     */
    val entity: Entity,
) {
    private val _sectionId = mutableVal(entity.sectionId.toInt())
    private val _section = mutableVal<SectionModel?>(null)
    private val _sectionInitialized = mutableVal(false)
    private val _position = mutableVal(vec3ToThree(entity.position))
    private val _worldPosition = mutableVal(_position.value)
    private val _rotation = mutableVal(vec3ToEuler(entity.rotation))
    private val _worldRotation = mutableVal(_rotation.value)

    val type: Type get() = entity.type

    val areaId: Int get() = entity.areaId

    val sectionId: Val<Int> = _sectionId

    val section: Val<SectionModel?> = _section
    val sectionInitialized: Val<Boolean> = _sectionInitialized

    /**
     * Section-relative position
     */
    val position: Val<Vector3> = _position

    val worldPosition: Val<Vector3> = _worldPosition

    /**
     * Section-relative rotation
     */
    val rotation: Val<Euler> = _rotation

    val worldRotation: Val<Euler> = _worldRotation

    val properties: ListVal<QuestEntityPropModel> = listVal(*Array(type.properties.size) {
        QuestEntityPropModel(this, type.properties[it])
    })

    open fun setSectionId(sectionId: Int) {
        entity.sectionId = sectionId.toShort()
        _sectionId.value = sectionId
    }

    fun initializeSection(section: SectionModel) {
        require(!sectionInitialized.value) {
            "Section is already initialized."
        }
        require(section.areaVariant.area.id == areaId) {
            "Section should lie within the entity's area."
        }

        setSectionId(section.id)

        _section.value = section

        // Update world position and rotation by calling setPosition and setRotation with the
        // current position and rotation.
        setPosition(position.value)
        setRotation(rotation.value)

        setSectionInitialized()
    }

    fun setSectionInitialized() {
        _sectionInitialized.value = true
    }

    /**
     * Will update the entity's relative transformation but keep its world transformation constant.
     */
    fun setSection(section: SectionModel) {
        require(section.areaVariant.area.id == areaId) {
            "Quest entities can't be moved across areas."
        }

        setSectionId(section.id)

        _section.value = section

        // Update relative position and rotation by calling setWorldPosition and setWorldRotation
        // with the current world position and rotation.
        setWorldPosition(worldPosition.value)
        setWorldRotation(worldRotation.value)
    }

    fun setPosition(pos: Vector3) {
        entity.setPosition(pos.x.toFloat(), pos.y.toFloat(), pos.z.toFloat())

        _position.value = pos

        val section = section.value

        _worldPosition.value =
            if (section == null) pos
            else pos.clone().applyEuler(section.rotation).add(section.position)
    }

    fun setWorldPosition(pos: Vector3) {
        val section = section.value

        val relPos =
            if (section == null) pos
            else (pos - section.position).applyEuler(section.inverseRotation)

        entity.setPosition(relPos.x.toFloat(), relPos.y.toFloat(), relPos.z.toFloat())

        _worldPosition.value = pos
        _position.value = relPos
    }

    fun setRotation(rot: Euler) {
        floorModEuler(rot)

        entity.setRotation(rot.x.toFloat(), rot.y.toFloat(), rot.z.toFloat())
        _rotation.value = rot

        val section = section.value

        if (section == null) {
            _worldRotation.value = rot
        } else {
            q1.setFromEuler(rot)
            q2.setFromEuler(section.rotation)
            q1 *= q2
            _worldRotation.value = floorModEuler(q1.toEuler())
        }
    }

    fun setWorldRotation(rot: Euler) {
        floorModEuler(rot)

        val section = section.value

        val relRot = if (section == null) {
            rot
        } else {
            q1.setFromEuler(rot)
            q2.setFromEuler(section.rotation)
            q2.invert()
            q1 *= q2
            floorModEuler(q1.toEuler())
        }

        entity.setRotation(relRot.x.toFloat(), relRot.y.toFloat(), relRot.z.toFloat())
        _worldRotation.value = rot
        _rotation.value = relRot
    }

    companion object {
        // These quaternions are used as temporary variables to avoid memory allocation.
        private val q1 = Quaternion()
        private val q2 = Quaternion()

        private fun floorModEuler(euler: Euler): Euler =
            euler.set(
                floorMod(euler.x, 2 * PI),
                floorMod(euler.y, 2 * PI),
                floorMod(euler.z, 2 * PI),
            )
    }
}
